Mathlib/Tactic/CancelDenoms/Core.lean

```diff
@@ -154,63 +154,61 @@ def synthesizeUsingNormNum (type : Q(Prop)) : MetaM Q($type) := do
     throwError "Could not prove {type} using norm_num. {e.toMessageData}"
 
 /--
-`mkProdPrf α sα v tr e` produces a proof of `v*e = e'`, where numeric denominators have been
+`mkProdPrf α sα v v' tr e` produces a proof of `v'*e = e'`, where numeric denominators have been
 canceled in `e'`, distributing `v` proportionally according to the tree `tr` computed
 by `findCancelFactor`.
+
+The `v'` argument is a numeral expression corresponding to `v`, which we need in order to state
+the return type accurately.
 -/
-partial def mkProdPrf {u : Level} (α : Q(Type u)) (sα : Q(Field $α)) (v : ℕ) (t : Tree ℕ)
-    (e : Q($α)) : MetaM Expr := do
+partial def mkProdPrf {u : Level} (α : Q(Type u)) (sα : Q(Field $α)) (v : ℕ) (v' : Q($α))
+    (t : Tree ℕ) (e : Q($α)) : MetaM ((e' : Q($α)) × Q($v' * $e = $e')) := do
```

Comment:
I suspect this could be made cleaner by pulling out this return type into its own `structure` declaration. Then you would not need to match on `_` so much, and maybe you could even add a constructor that takes `e'` implicitly.

Reply:
I don't really see how the `_` matching would be avoided; I guess making it an explicit argument would make the `_` invisible, but morally it would still be there.

Reply:
I was imagining something like

```lean
/-- Return value of the `mkProdPrf` function. -/
structure MkProdPrfResult {u : Level} {α : Q(Type u)} (sα : Q(Field $α)) (v' e : Q($α)) where
  {e' : Q($α)}
  pf : Q($v' * $e = $e')
```

Then you could do
```lean
...
  | .node _ lhs rhs, ~q($e1 + $e2) => do
    let ⟨v1⟩ ← mkProdPrf α sα v v' lhs e1
    let ⟨v2⟩ ← mkProdPrf α sα v v' rhs e2
    return ⟨q(CancelDenoms.add_subst $v1 $v2)⟩
...
```

This seems more readable to me, but maybe I'm overthinking it.

Reply:
I do agree that a `Result` structure like that is useful, I've used it as well for my own (experimental) tactic implementations. You could even reuse `Simp.Result` if you want to make it easy to adapt this into a `simproc`.

On the other hand, I consistent use of `Result` structures would require much more work than what this PR is intended to do. So I'll let Eric decide if this is worth it.

Reply:
I adjusted to use a structure; though rather than removing the `_`s, I opted to make their values explicit to make it easier to follow what the tactic is doing.

Reply:
@dwrensha, can you check you're happy with this?

bors d=@dwrensha